Use nested namespace definitions

barcode opened the pull request:
Replace `namespace a { namespace b { ... }}` with nested namespaces.

steven-varga commented:
Hi,
yes you could do that,  what is the value you are adding? -- Background: I year ago when I approached the HDFGroup with the idea of a modern C++ implementation I sketched it in c++17, then under request I lowered it to c++14 and then tried c++11. This has to do with various compilers and linear algebra systems used in super computing. For now I put aside c++11. The work is not justified. If you wish to contribute, it is fine to do it in c++17.

The current issues AFAIK:

1. variable length massive string  stored in heap structure -- there is no profiling done on terra byte level, we just don't know its performance characteristics of this method. 
2. fixed length strings performance: space and throughput -- is it possible to do better?
3. pending framework for multi threaded filtering chain: this spring I did the study then sketched  out the idea for multi threaded compression/filter chain. This work requires understanding of processor caching and memory management. -- finicky but you may be interested. The goal is to take a server class processor to its limits. 
4. connecting to other systems: protocol buffers, thrift, ... provides easy way to connect to other systems, adding features to persist these streams can make HDF5 as a format more widespread.
5. adding more BLAS/LAPACK based linear algebra systems: the mechanism is established, so the majority of the work is to do a thorough research what I missed, and if worth adding them.
6. PBLAS/scaLAPACK based systems require efficient data distribution routines, this one requires supercomputer/properly tuned cluster access, and understanding of MPI patterns.
7. documentation: doxygen didn't quite work out for template meta-programming, an alternative solution may work better?

Let me know if you have interest in working on any of these ideas or briefly outline an idea of your own. 

best wishes: steven

barcode commented:
I did this to improve readability while reading the code.
And according to your answer in #9, you:

 > will yank all awkward c++11 -ishness and even 14-ness from the codebase, and go with the more powerful c++17 -- the way I originally intended this to be.

Hence i thought i would help with this (since i was doing this anyway).

In the meantime, i have converted more code to `c++17` and added some code documentation.
I will add PRs for that.

> The current issues AFAIK:
> ...

None of these issues really come up in my work. I am using h5cpp primarily to serialize data in a fashion which is accessible to other languages (e.g.: Python). Most of these data is for machine learning at a small / medium scale (few 100GB of data).

Issue 7: I know [Eigen3](http://eigen.tuxfamily.org/dox/) and [CGAL](https://doc.cgal.org/latest/Manual/index.html) use doxygen for their documentation and both have a lot of template meta-programing.
Hence I assume it is possible to document this library with doxygen.

> Let me know if you have interest in working on any of these ideas 

Currently i do not really have the use / test-cases or the knowledge on how to deal with the issues.

> or briefly outline an idea of your own.

Mainly I am doing convenience / quality of life features on the interface layer.
I want to make it easier to write custom structs to hdf5. These structs might contain statically sized `Eigen::Matrix` objects. I know there is the `h5cpp-compiler`, but i do not really like to use a different tool if i can express this directly in code. Hence i build a macro similar to [BOOST_FUSION_ADAPT_STRUCT](https://www.boost.org/doc/libs/1_70_0/libs/fusion/doc/html/fusion/adapted/adapt_struct.html) (the boost macro did not capture all necessary information). This macro generates code necessary to write the structs with h5cpp and can be put after an include to `h5cpp/all`. 

    #include <h5cpp/all>
    struct s { int a; int b; };
    ADAPT(s, a, b) // the name will change
    //now h5::read<std::vector<s>>(...) works

As a side product, you can directly read/write types similar to `std::vector<Eigen::Matrix4f>` as compound structs.
At the moment, all of this works (except directly writing fixed size Eigen matrices. Currently i am doing some cleaning up and improving.
